### ISSUE TITLE
KCT changes

### ISSUE THREAD

**Issue by NathanKell** [KSP-RO/RP-0#965]
We've talked about these in discord. But I wanted to collect the changes we wanted to make:

* Ability to *decrease* upgrade points in a given area (no refund, just for lowering maintenance costs).
~~* Handling rollout better (making some of it an integration cost, not recharging that cost/time on rollback/re-launch)~~
~~* Have KCT check for the ability to launch a craft with unlocked parts~~
~~* Configurable cost-per-click for rush build~~
~~* Ability to choose pad level to build to when building a new pad~~
~~* Instead of 1pt per node unlocked, 1pt per (x) science points gained.~~
~~* Allow gating of facility upgrades by techs (prevent upgrading a building if you lack a tech)~~
~~* Prevent rolling out to a pad that is currently being upgraded~~
~~* Show total upgrade points invested in each VAB tier / R&D (best on the tab where the rate is shown, not just Upgrades)~~
~~* Training: Allow training on a part being researched~~

**Comment by pap1723** [comment]
Configurable cost per click for rush build is implemented. I wrote that today and added it to our RP0 KCT Preset.

**Comment by pap1723** [comment]
After digging in the code today I believe that the last one would be relatively easy as well. We could hook it to the "Launch" button and it charges automatically on the click. This would also stop people from automatically building 2 of everything like we all do now.

However, that would not be entirely realistic. Probably a much better way would be to separate out the cost over the length of the build like rollout currently works. So it would charge over time as it is being "integrated, tested, etc".

**Comment by lpgagnon** [comment]
QoL improvement: make KCT display somewhere how many upgrade points were put into the VAB, how many into R&D. (as opposed to what we have now in the Upgrade window, a bare total points).

For me, it would make it just a bit easier to stick to a particular ratio.

(for bonus points: show this, as well as the BP/s and sci/yr, in the main kct window instead of hiding it in the Upgrade window)

**Comment by Bornholio** [comment]
^^^ this first item Also can we have zero pads in a new area of Switcher until its paid for in VAB Points or Money. Ignore me i'm drunk

**Comment by lpgagnon** [comment]
"Ah, but that's the point--integration happens during "rollout". The time spent "building" the rocket is the time spent for the stages to be in construction / get delivered. Once it is "built", then it goes into the VAB for integration. The time spent actually physically rolling out to the pad is only about a third of the "rollout" time KCT talks about (first third is integration, last third is on-the-pad work. Although it's probably more 50/20/30 split.)"

The way it's implemented really doesn't make much sense, though. If integration time is part of rollout time, why does upgrading the VAB let me build 2 rockets in parallel? Why do I need multiple launchpads in order to integrate multiple rockets at the same time?

**Comment by NathanKell** [comment]
A suggestion by IIRC @Bornholio : don't have free pads everywhere. You only get a pad at the site you actually use, and switching to another KSCSwitcher site doesn't give you another pad.

**Comment by lpgagnon** [comment]
“You only get a pad at the site you actually use” one problem is right now there’s no mechanism to specify up front which site you want to use. Alternative: start with no pad anywhere?

**Comment by NathanKell** [comment]
When you first start building an LV, go through all other sites and remove their pads.

**Comment by NathanKell** [comment]
Oh and FIX THE ROLLOUT COST DISPLAY BUG. Pls.

**Comment by NathanKell** [comment]
Checked, and gating facility upgrades by tech looks totally doable.

**Comment by ajvaughan** [comment]
[From discord] If you are going to prevent rolling out to a pad that is being upgraded, can you add a warning to the pad upgrade confirmation dialog.  Maybe : "Upgrade of Pad xxx to  xxx will take xxx days.\n  During that time you won't be able to use the pad".  
(I can't remember whether there currently is a confirmation dialog, but I think there should be a warning about the pad being unusable for xxx days, otherwise someone might screw themselves by failing a few contracts.

**Comment by NathanKell** [comment]
That, and the block-launching-to-upgrading pads thing, is done.
@siimav is adding the "create new pad at specified level" thing.

**Comment by leudaimon** [comment]
Is building a pad directly at a higher level cheaper and/or faster than doing it stepwise? I guess it should be to make it worth it.

**Comment by NathanKell** [comment]
Adding one more that we talked about on discord but I forgot to include originally: allowing training once a node is being researched, rather than waiting until the part has been unlocked. The easier way to do this would just be to include all parts in researchable nodes.

The downside is the course menu will get quite cluttered, because right now it only shows parts you have affirmatively unlocked (and paid an entry cost for, so you won't have done it much).

**Comment by siimav** [comment]
By now the only things not implemented from the original list are:
* Ability to decrease upgrade points
* Handling rollout better

Also the suggestion to remove free pads (and runways?) from KSCs other than the main one.

**Comment by leudaimon** [comment]
I feel the ability to decrease upgrade points is not very much high priority...
About the removal of all pads/runways, how would we handle that, without forcing people to use KSC? Would it be easier to remove them all by default and somehow build one only in 1 site the player chooses?

**Comment by pap1723** [comment]
I think having free runways and pads at the other sites is alright. They are level 1 only, so the sounding rockets that are typically launched from there could realistically usually be launched in the middle of a field. (I know I am exaggerating). They very quickly lose their effectiveness.

I also think that decreasing upgrade points is something that can wait. I see the benefit, but I don't know how many people will be using it. The big one is still the only one that is left. That will be for the future.

**Comment by pap1723** [comment]
Updated first post

**Comment by leudaimon** [comment]
About the lvl 1 pads everywhere, my concern about keeping them is biome fishing with sounding rockets. But I guess this is a minor issue out of a RIS game.

**Comment by smartdummies** [comment]
If the ability to decrease the build points allows me to remove the initial point in the SPH I would use it for sure. I would even appreciate the ability to remove my runway at my initial site. I know the maintenance is small, but early game I would take every bit of help I can get. A thought for this is to remove all the launch pads and runways and require the user to construct the initial one at their chosen starting location, starting funds increased appropriately to cover. The build time for this initial build should be fairly quick so that it can be completed in conjunction with the first sounding rocket. This should also prevent the launching of a "fake" quick first sounding rocket (say Areobee core and sep motor) to get the "First Launch" contract completed unrealistically early.

**Comment by siimav** [comment]
Splitting build and integration has now been mostly implemented on SepIntegration branches for both KCT and RP-1. I decided to create 2 new formulas for the KCT preset: `IntegrationTimeFormula` and `IntegrationCostFormula`. Currently the integration cost is deducted as soon as the vessel is added to the build queue. To reduce the changes needed in KCT code, Integration time is added on top of the build time and is technically not handled as a separate step between build and rollout.
[Here's the spreadsheet](https://docs.google.com/spreadsheets/d/1EnA37j5sno1ffn_-pIhICkqSMccn3aG_rUFPyxqFgcU) that I used for comparing launch costs and times to come up with a formula that minimizes the impact to early/mid-game balancing.
